### ISSUE TITLE
Controller ID Mismatch

### ISSUE THREAD

**Issue by avencherus** [godotengine/godot#12643]
On Windows 10, with stable 2.1.4 build, using PS4 Controllers.

**Also still in 3.1 alpha**

Example Project: [Joypad Issue.zip](https://github.com/godotengine/godot/files/2356061/Joypad.Issue.zip)

When using multiple controllers, the IDs get mixed up if you plug and unplug them in a particular order.

This problem shows up in the **Input** signal **joy_connection_changed**, and will also manifest oddly in the **InputEvent.device** property.

To recreate signal problem:
* Have two controllers plugged in, figure out which one has ID or **index = 0**.
* Unplug the controller with ID zero.  The signal code will report an **index = 1**.
* Now unplug the other controller, it's ID has shifted to 0.  The signal will report this index as zero.

To recreate **InputEvent** problem:
* Locate the controller that has **index = 1**.  I will label it Controller 2, and the one with ID zero is Controller 1.
* Unplug Controller 2.
* Unplug Controller 1.
* Plug Controller 2 back in.
* Plug Controller 1 back in.
* Start hitting buttons on both controller and observe the **InputEvent.device**.
* Controller 1 will return nothing, and Controller 2 will return both 0 and 1.

If there might be something dumb I'm doing to cause this, please let me know.

**Comment by avencherus** [comment]
@Hinsbart Thanks for looking into it.  X)

**Comment by avencherus** [comment]
It seems almost fixed in 2.1.5 beta and 3.  The indexes no longer appear to get completely lost, crossed up, and stacked anymore, but when the first controller is unplugged and plugged back in, the second controller will increment its index and then send a **joy_connection_changed** signal with the values of _true and the new index_.

**Unplug controller 1** 
index = 0, connection = false

**Plug in controller 1**
index = 0, connection = true
index = 2, connection = true   <-- This is the second controller.

I imagine it isn't intentional to have the index increment like this, since it isn't reporting its disconnection.

Can't easily rely on the signal, but at least in the current state, with variables and some extra logic in the script it can be tracked and worked around.  At least for two controllers.  Not sure about the pattern of the index shifting with 3 or more.

**Comment by akien-mga** [comment]
Any idea @Hinsbart?

**Comment by avencherus** [comment]
@akien-mga @Hinsbart 

Still present in beta 2, with an additional wrinkle.  When the controller with id 0 gets unplugged.  The controllers with the 1+ id, will stop giving signals when it plugs and unplugs.  It will resume giving signals only when the previous controller is plugged back in.  That's also when it's id gets incremented.

**Comment by reduz** [comment]
@Hinsbart any progress on this?

**Comment by avencherus** [comment]
@reduz Hey Juan, from looking at it yesterday, I can say it is still in 3.1.  It's a problem on Windows, but seemed to be working fine on Ubuntu at least.

I gandered at the joypad code, and I don't know enough about the libraries to follow the logic.  But it seems like it is some sort of index leak for the device IDs.  As if maybe when the first controller reconnects the second controller is also getting reconnected, but never getting disconnected and relinquishing its previous ID.

If you unplug and replug the first controller in and out, while a second stays connected, the ID array keeps growing by one, in relation to the 2nd joypad.  The first index (0) does clean itself up.

The array looks like this as you disconnect and reconnect joypad 1.  It uses ID 0.

> [0, 1]
> [1]
> [0, 1, 2]
> [1, 2]
> [0, 1, 2, 3]

On and on it goes.  I made some crazy work around for it that can support at least for 2 joypads, by taking the first and last IDs, and tracking the increment if there is a 0 or not in the first position.  It at least prevents fatal crashes, but the second controller does become unresponsive in one specific condition.  So it is not ideal, and quite a bit of extra code.

I'm porting this hack over to my new project today, and I was considering posting an update example project for it soon.
